esrally/telemetry.py

```diff
@@ -1408,6 +1409,199 @@ def record(self):
             self.metrics_store.put_doc(doc, level=MetaInfoScope.cluster, meta_data=data_stream_metadata)
 
 
+class IngestPipelineStats(InternalTelemetryDevice):
+    command = "ingest-pipeline-stats"
+    human_name = "Ingest Pipeline Stats"
+    help = "Reports Ingest Pipeline stats at the end of the benchmark."
+
+    def __init__(self, clients, metrics_store):
+        super().__init__()
+        self.logger = logging.getLogger(__name__)
+        self.clients = clients
+        self.metrics_store = metrics_store
+        self.start_stats = {}
+        self.specified_cluster_names = self.clients.keys()
+
+        self.ingest_pipeline_cluster_count = 0
+        self.ingest_pipeline_cluster_time = 0
+        self.ingest_pipeline_cluster_failed = 0
+
+    def on_benchmark_start(self):
+        self.logger.info("Gathering Ingest Pipeline stats at benchmark start")
+        self.start_stats = self.get_ingest_pipeline_stats()
+
+    def on_benchmark_stop(self):
+        self.logger.info("Gathering Ingest Pipeline stats at benchmark end")
+        end_stats = self.get_ingest_pipeline_stats()
+
+        # The nesting level is ok here given the structure of the Ingest Pipeline stats
+        # pylint: disable=too-many-nested-blocks
+        for cluster_name, node in end_stats.items():
+            if cluster_name not in self.start_stats:
+                self.logger.warning(
+                    "Cannot determine Ingest Pipeline stats for %s (cluster stats weren't collected at the start of the benchmark).",
+                    cluster_name,
+                )
```

Comment:
Why is this only a warning and not an error, by the way? This should not happen, right? (Same for the others.)

Reply:
I put it as warning on the assumption that if you were analysing multiple clusters, we may not want to fail the benchmark because we failed to report ingest pipeline stats for one of those clusters .

That said, it probably does make sense to error out at lower levels (node, pipeline, processor). What do you think?

Reply:
I'm torn here. It feels like all logs should be at the same log level, and it's not fatal, so I think we can leave it like this